File: sdks/go/cmd/elydora/plugins/runtime_paths.go
package plugins

import (
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"runtime"
	"strings"
)

var windowsDeviceName = regexp.MustCompile(
	`(?i)^(?:con|prn|aux|nul|com[1-9¹²³]|lpt[1-9¹²³])(?:\.|$)`,
)

// AgentRuntimeRoot returns the local root for all Elydora agent runtimes.
func AgentRuntimeRoot() (string, error) {
	home, err := os.UserHomeDir()
	if err != nil {
		return "", fmt.Errorf("resolve home directory: %w", err)
	}
	root, err := filepath.Abs(filepath.Join(home, ".elydora"))
	if err != nil {
		return "", fmt.Errorf("resolve Elydora runtime root: %w", err)
	}
	return root, nil
}

// ResolveAgentRuntimeDirectory confines an agent ID to one portable path segment.
func ResolveAgentRuntimeDirectory(agentID string) (string, error) {
	if invalidAgentDirectoryName(agentID) {
		return "", fmt.Errorf(
			"invalid agent ID for local storage: %q must be a single non-empty path segment with a portable file name",
			agentID,
		)
	}
	root, err := AgentRuntimeRoot()
	if err != nil {
		return "", err
	}
	candidate := filepath.Join(root, agentID)
	relative, err := filepath.Rel(root, candidate)
	if err != nil {
		return "", fmt.Errorf("resolve agent runtime path for %q: %w", agentID, err)
	}
	if relative == "." || relative == ".." || filepath.IsAbs(relative) ||
		strings.HasPrefix(relative, ".."+string(os.PathSeparator)) ||
		strings.ContainsRune(relative, os.PathSeparator) {
		return "", fmt.Errorf("agent ID escapes the local storage directory: %q", agentID)
	}
	if _, err := RequirePhysicalDirectory(root); err != nil {
		return "", err
	}
	if _, err := RequirePhysicalDirectory(candidate); err != nil {
		return "", err
	}
	return candidate, nil
}

func invalidAgentDirectoryName(agentID string) bool {
	if agentID == "" || agentID == "." || agentID == ".." ||
		strings.ContainsAny(agentID, `<>:"/\|?*`) ||
		strings.HasPrefix(agentID, " ") || strings.HasSuffix(agentID, ".") ||
		strings.HasSuffix(agentID, " ") ||
		windowsDeviceName.MatchString(agentID) {
		return true
	}
	for _, character := range agentID {
		if character <= 31 {
			return true
		}
	}
	return false
}

// EnsurePrivateDirectory creates an owner-only physical directory.
func EnsurePrivateDirectory(path string) error {
	if err := os.MkdirAll(path, 0700); err != nil {
		return fmt.Errorf("create private directory %s: %w", path, err)
	}
	info, err := os.Lstat(path)
	if err != nil {
		return fmt.Errorf("inspect private directory %s: %w", path, err)
	}
	if info.Mode()&os.ModeSymlink != 0 || !info.IsDir() {
		return fmt.Errorf("private directory path is not a physical directory: %s", path)
	}
	if runtime.GOOS != "windows" {
		if err := os.Chmod(path, 0700); err != nil {
			return fmt.Errorf("restrict private directory %s: %w", path, err)
		}
	}
	return nil
}

// PrepareAgentRuntimeDirectory validates and creates the private runtime path.
func PrepareAgentRuntimeDirectory(agentID string) (string, error) {
	root, err := AgentRuntimeRoot()
	if err != nil {
		return "", err
	}
	agentDirectory, err := ResolveAgentRuntimeDirectory(agentID)
	if err != nil {
		return "", err
	}
	if err := EnsurePrivateDirectory(root); err != nil {
		return "", err
	}
	if err := EnsurePrivateDirectory(agentDirectory); err != nil {
		return "", err
	}
	return agentDirectory, nil
}

// RequirePhysicalDirectory reports whether a path is an existing physical directory.
func RequirePhysicalDirectory(path string) (bool, error) {
	info, err := os.Lstat(path)
	if os.IsNotExist(err) {
		return false, nil
	}
	if err != nil {
		return false, fmt.Errorf("inspect agent runtime directory %s: %w", path, err)
	}
	if info.Mode()&os.ModeSymlink != 0 || !info.IsDir() {
		return false, fmt.Errorf("agent runtime path is not a physical directory: %s", path)
	}
	return true, nil
}

// RequirePhysicalFile reports whether a path is an existing physical regular file.
func RequirePhysicalFile(path string) (bool, error) {
	info, err := os.Lstat(path)
	if os.IsNotExist(err) {
		return false, nil
	}
	if err != nil {
		return false, fmt.Errorf("inspect agent runtime file %s: %w", path, err)
	}
	if info.Mode()&os.ModeSymlink != 0 || !info.Mode().IsRegular() {
		return false, fmt.Errorf("agent runtime config is not a physical file: %s", path)
	}
	return true, nil
}

File: sdks/node/src/runtime-paths.ts
import fsp from 'node:fs/promises';
import path from 'node:path';

export function isMissingPath(error: unknown): boolean {
  return error instanceof Error && 'code' in error && error.code === 'ENOENT';
}

export async function ensurePrivateDirectory(directory: string): Promise<void> {
  await fsp.mkdir(directory, { recursive: true, mode: 0o700 });
  const metadata = await fsp.lstat(directory);
  if (!metadata.isDirectory() || metadata.isSymbolicLink()) {
    throw new Error(`Private directory path is not a physical directory: ${directory}`);
  }
  if (process.platform !== 'win32') {
    await fsp.chmod(directory, 0o700);
  }
}

export async function requirePhysicalDirectory(directory: string): Promise<boolean> {
  let metadata;
  try {
    metadata = await fsp.lstat(directory);
  } catch (error) {
    if (isMissingPath(error)) return false;
    throw new Error(`Could not inspect agent runtime directory: ${directory}`, { cause: error });
  }
  if (!metadata.isDirectory() || metadata.isSymbolicLink()) {
    throw new Error(`Agent runtime path is not a physical directory: ${directory}`);
  }
  return true;
}

export async function requirePhysicalFile(filePath: string): Promise<boolean> {
  let metadata;
  try {
    metadata = await fsp.lstat(filePath);
  } catch (error) {
    if (isMissingPath(error)) return false;
    throw new Error(`Could not inspect agent runtime file: ${filePath}`, { cause: error });
  }
  if (!metadata.isFile() || metadata.isSymbolicLink()) {
    throw new Error(`Agent runtime config is not a physical file: ${filePath}`);
  }
  return true;
}

export function resolvePrivateChildDirectory(root: string, childName: string): string {
  const isWindowsDeviceName = /^(?:con|prn|aux|nul|com[1-9¹²³]|lpt[1-9¹²³])(?:\.|$)/i.test(childName);
  if (
    childName.length === 0
    || /[<>:"|?*\u0000-\u001f]/.test(childName)
    || childName.includes('/')
    || childName.includes('\\')
    || childName === '.'
    || childName === '..'
    || childName.startsWith(' ')
    || childName.endsWith('.')
    || childName.endsWith(' ')
    || isWindowsDeviceName
  ) {
    throw new Error(`Invalid agent ID for local storage: ${JSON.stringify(childName)}`);
  }

  const resolvedRoot = path.resolve(root);
  const candidate = path.resolve(resolvedRoot, childName);
  const relative = path.relative(resolvedRoot, candidate);
  if (
    relative.length === 0
    || path.isAbsolute(relative)
    || relative === '..'
    || relative.startsWith(`..${path.sep}`)
    || relative.includes(path.sep)
  ) {
    throw new Error(`Agent ID escapes the local storage directory: ${JSON.stringify(childName)}`);
  }
  return candidate;
}
